File: 239-sliding-window-maximum/sliding-window-maximum.py
class Solution(object):
    def maxSlidingWindow(self, nums, k):
     ans=[]
     dq=deque()
     n=len(nums)
     for i in range(k):
      while dq and nums[i]>=nums[dq[-1]]:
       dq.pop()
      dq.append(i)
      
     for i in range(k,n):
      ans.append(nums[dq[0]])
      #push incoming ele
      while dq and nums[i]>=nums[dq[-1]]:
        dq.pop()
      dq.append(i)
      #remove ele from previous window
      while dq and dq[0]<=i-k:
       dq.popleft()
     
    #  add max of last window
     ans.append(nums[dq[0]])
     return ans

File: 239-sliding-window-maximum/sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        priority_queue<pair<int,int>>pq;
        for(int i=0;i<k;i++){
            pq.push({nums[i],i});
        }
        int n=nums.size();
        vector<int>ans;
        ans.push_back(pq.top().first);

        for(int i=k;i<n;i++){
            pq.push({nums[i],i});

            while(i-k>=(pq.top().second)){
                pq.pop();
            }

            ans.push_back(pq.top().first);
        }
        return ans;
    }
};